### PR TITLE
Define eq_dyn_scalar API

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Working on this in relation to #984 and #1068 with the end goal being to finalize how we want `eq_dyn_scalar` to work.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
@alamb ive started the work on defining the `eq_dyn_scalar` api.  ive tried to combine some of the different points we discussed / you proposed over the course of #984. do you think this is going in the right direction?

File: arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -2522,4 +2603,19 @@ mod tests {
...
+    #[test]
+    fn test_eq_dyn_scalar_with_dict() {
+        let key_builder = PrimitiveBuilder::<UInt8Type>::new(3);
+        let value_builder = PrimitiveBuilder::<UInt8Type>::new(2);
+        let mut builder = PrimitiveDictionaryBuilder::new(key_builder, value_builder);
+        builder.append(123).unwrap();
+        builder.append_null().unwrap();
+        builder.append(223).unwrap();
+        let array = builder.finish();
+        let a_eq = eq_dyn_scalar(&array, 123).unwrap();
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:2615]: looking good to me 👍

File: arrow/src/compute/kernels/comparison.rs [hunk 2]
```diff
@@ -898,6 +898,87 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+{
+    match left.data_type() {
+        DataType::UInt8 => {
+            let right: u8 = right.try_into().map_err(|_| {
+                ArrowError::ComputeError(format!(
+                    "Can not convert {:?} to u8 for comparison with UInt8Array",
+                    right
+                ))
+            })?;
+            eq_scalar::<UInt8Type>(as_primitive_array::<UInt8Type>(left), right)
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:922]: 👍

**Review by alamb — COMMENTED** [review]
looks good @matthewmturner  

One challenge is going to be making code that is maintainable and not just swaths of copy/paste (but different type code). It will be a fun exercise

**Comment by alamb** [comment]
cc @shepmaster , @jorgecarleitao  and @alippai

**Comment by matthewmturner** [comment]
@alamb yes agree. That's what I'm going to work on next, the first commit was basically just what you said (copy/paste) with minimal cleaning to show the high level api. Now going to try making macros that can be used for all dyn_scalar kernels. 

Just to confirm, we're still expecting type annotation on the scalar right? I.e 123u8

**Comment by alamb** [comment]
> Just to confirm, we're still expecting type annotation on the scalar right? I.e 123u8

I am hoping that we can avoid those annotations, to be honest

**Comment by matthewmturner** [comment]
@alamb i think i might be getting close.  I used your `TryInto<i128>` idea to enable using the kernel without type annotations.

Not sure why im hitting a recursion limit error though - will need to do some more work on that.

Do you think this is going in the right direction?

File: arrow/src/compute/kernels/comparison.rs [hunk 3]
```diff
@@ -898,6 +898,126 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
     compare_op_scalar!(left, right, |a, b| a >= b)
 }
 
+macro_rules! dyn_cmp_scalar {
+    ($LEFT: expr, $RIGHT: expr, $T: ident, $OP: ident, $TT: tt) => {{
```
Comment (shepmaster) [inline arrow/src/compute/kernels/comparison.rs:902]: Should `$T` and `$TT` be [`ty`](https://doc.rust-lang.org/stable/reference/macros-by-example.html#metavariables)?

File: arrow/src/compute/kernels/comparison.rs [hunk 4]
```diff
@@ -898,6 +898,126 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+                type_name::<$T>()
+            ))
+        })?;
+        let right = $RIGHT.as_any().downcast_ref::<$T>().ok_or_else(|| {
+            ArrowError::CastError(format!(
+                "Right array cannot be cast to {}",
+                type_name::<$T>(),
+            ))
+        })?;
+        $OP::<$TT>(left, right)
```
Comment (shepmaster) [inline arrow/src/compute/kernels/comparison.rs:915]: `$OP::<$TT>` could probably be fused as one `expr` macro argument

File: arrow/src/compute/kernels/comparison.rs [hunk 5]
```diff
@@ -898,6 +898,126 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+            (t1, t2) => Err(ArrowError::CastError(format!(
+                "Cannot compare an array with a scalar of different type ({} and {})",
+                t1, t2
+            ))),
+        }
+    }};
+}
+
+/// Perform `left == right` operation on an array and a numeric scalar
+/// value. Supports PrimtiveArrays, and DictionaryArrays that have primitive values
```
Comment (shepmaster) [inline arrow/src/compute/kernels/comparison.rs:985]:
```suggestion
/// value. Supports PrimitiveArrays, and DictionaryArrays that have primitive values
```

File: arrow/src/compute/kernels/comparison.rs [hunk 6]
```diff
@@ -898,6 +898,126 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+/// TODO add example
+///
+fn unpack_dict_comparison<K>(
+    left: &DictionaryArray<K>,
+    dict_comparison: BooleanArray,
+) -> Result<BooleanArray>
+where
+    K: ArrowNumericType,
+{
+    assert_eq!(dict_comparison.len(), left.values().len());
```
Comment (shepmaster) [inline arrow/src/compute/kernels/comparison.rs:1004]: Why an assertion as opposed to an error or a "no this does not match?"

File: arrow/src/compute/kernels/comparison.rs [hunk 7]
```diff
@@ -898,6 +898,126 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+                type_name::<$T>(),
+            ))
+        })?;
+        $OP::<$TT>(left, right)
+    }};
+}
+
+macro_rules! dyn_compare_scalar {
+    ($LEFT: expr, $RIGHT: expr, $OP: ident) => {{
+        let right = $RIGHT.try_into().map_err(|_| {
```
Comment (shepmaster) [inline arrow/src/compute/kernels/comparison.rs:921]: I'm missing where this `right` value is used...

File: arrow/src/compute/kernels/comparison.rs [hunk 8]
```diff
@@ -898,6 +898,126 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+    }};
+}
+
+/// Perform `left == right` operation on an array and a numeric scalar
+/// value. Supports PrimtiveArrays, and DictionaryArrays that have primitive values
+pub fn eq_dyn_scalar<T>(left: &dyn Array, right: T) -> Result<BooleanArray>
+where
+    T: IntoArrowNumericType + TryInto<i128> + Copy + std::fmt::Debug,
+{
+    dyn_compare_scalar!(left, right, eq_scalar)
```
Comment (shepmaster) [inline arrow/src/compute/kernels/comparison.rs:990]: I admit this is a drive-by review, but I'm not seeing the benefit of the macros here yet. They don't do any repetition reduction. It looks like `dyn_compare_scalar` could be inlined and `dyn_cmp_scalar` could be a regular function.

File: arrow/src/compute/kernels/comparison.rs (hunk 7 above)
Comment (matthewmturner) [reply arrow/src/compute/kernels/comparison.rs:921]: I just hadn't yet updated where $RIGHT is used to use right instead

**Comment by matthewmturner** [comment]
Ive made a number of updates to simplify this, such as de-macroing a level, and have finally made our test pass.  Of course there would need to be considerable cleanup to his but I think that this could demonstrate an approach to making this work.  To be fair, ive thought ive been close about 5 times now and have been wrong each time so definitely happy to get more insight from others.

@alamb as always your feedback greatly appreciated.

@shepmaster thank you for the review before.  ive cleaned up based on some of your feedback with the goal of trying to demonstrate in simpler terms (i.e. less macros) what were trying to do.  im still quite new to rust so im not sure how to manage inlining `dyn_compare_scalar`.  ive started to read up on it but will definitely take more time to get an understanding.  If you could provide any additional color on that it would be greatly appreciated, else I can follow up once i have a better understanding).  also would be interested in what @alamb has to say about it.

**Comment by matthewmturner** [comment]
also as a note, i added the `IntoArrowNumericType` just for testing purposes here, of course it wouldnt actually be added there.  I also wasnt sure how to go about accessing the associated type `Arrow` (i.e. in a match) so added a method to get the relevant arrow type.

File: arrow/src/compute/kernels/comparison.rs (hunk 6 above)
Comment (alamb) [reply arrow/src/compute/kernels/comparison.rs:1103]: I think it is an invariant (namely that `left` is the dictionary and `dict_comparison` is the result of comparing those values). 

Perhaps we could rename the `left` parameter to `dict` to make this clearer

File: arrow/src/compute/kernels/comparison.rs [hunk 9]
```diff
@@ -898,6 +899,224 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+            (DataType::Int8, DataType::Int8) => {
+                let right: i8 = right.try_into().map_err(|_| {
+                    ArrowError::ComputeError(String::from(
+                        "Can not convert scalar to i128",
+                    ))
+                })?;
+                let left =
+                    $LEFT.as_any().downcast_ref::<Int8Array>().ok_or_else(|| {
+                        ArrowError::CastError(String::from("Left array cannot be cast"))
+                    })?;
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:920]: I think you can use https://docs.rs/arrow/6.4.0/arrow/array/fn.as_primitive_array.html to simplify this

So something like

```suggestion
                    as_primitive_array::<Int8Array>($LEFT)?;
```

File: arrow/src/compute/kernels/comparison.rs [hunk 10]
```diff
@@ -898,6 +899,224 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+where
+    T: IntoArrowNumericType + TryInto<i128> + Copy + std::fmt::Debug,
+{
+    let result = match left.data_type() {
+        DataType::Dictionary(key_type, _) => match key_type.as_ref() {
+            DataType::UInt8 => {
+                let left = left
+                    .as_any()
+                    .downcast_ref::<DictionaryArray<UInt8Type>>()
+                    .unwrap();
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:1047]: You might be able to use https://docs.rs/arrow/6.4.0/arrow/array/fn.as_dictionary_array.html here too:

```suggestion
                let left = as_dictionary_array<DictionaryArray<UInt8Type>>()?;
```

File: arrow/src/compute/kernels/comparison.rs [hunk 11]
```diff
@@ -898,6 +899,224 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+    fn get_arrow_type(&self) -> &DataType {
+        &DataType::Int32
+    }
+}
+
+/// Perform `left == right` operation on an array and a numeric scalar
+/// value. Supports PrimitiveArrays, and DictionaryArrays that have primitive values
+pub fn eq_dyn_scalar<T>(left: &dyn Array, right: T) -> Result<BooleanArray>
+where
+    T: IntoArrowNumericType + TryInto<i128> + Copy + std::fmt::Debug,
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:1039]: If `T` is going to have `TryInto<i128>` , I wonder if we still `IntoArrowNumericType` at all? I think it may not be necessary any more. 

My thinking is that since `dyn_compare_scalar` converts `right` into `i128` immediately there are then conversion rules back to all of the primitive types needed to call `eq_scalar`

**Review by alamb — COMMENTED** [review]
Looking really close @matthewmturner  -- 🙏  -- thank you. I do wonder about having both `IntoNumericType` and  `Into<i128>` I feel like like they are redundant somehow

**Comment by matthewmturner** [comment]
@alamb to confirm, are you expecting this kernel to handle boolean and utf8 comparisons as well?

File: arrow/src/compute/kernels/comparison.rs (hunk 9 above)
Comment (matthewmturner) [reply arrow/src/compute/kernels/comparison.rs:921]: I think these functions require an `Arc<dyn Array>` /`ArrayRef` which would mean we need the end user to do that before passing the array to the kernel or do you think its okay to add the `Arc` in the kernel?

**Comment by matthewmturner** [comment]
@alamb to confirm, we arent handling floats with this kernel right? My understanding is our current `T` wouldnt work with that.

**Comment by matthewmturner** [comment]
Ive added `eq_dyn_utf8_scalar` as well

File: arrow/src/compute/kernels/comparison.rs [hunk 12]
```diff
@@ -898,6 +900,261 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+}
+
+/// Perform `left == right` operation on an array and a numeric scalar
+/// value. Supports StringArrays, and DictionaryArrays that have string values
+pub fn eq_dyn_utf8_scalar(left: Arc<dyn Array>, right: &str) -> Result<BooleanArray> {
+    match left.data_type() {
+        DataType::Dictionary(key_type, _) => {
+            return dyn_compare_utf8_scalar!(&left, right, key_type, eq_utf8_scalar);
+        }
+        _ => {
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:1123]: this probably wants to match on the `DataType::Utf8` and `DataType::LargeUtf8` but otherwise looks good to me

File: arrow/src/compute/kernels/comparison.rs [hunk 13]
```diff
@@ -2522,4 +2779,60 @@ mod tests {
...
+        assert_eq!(
+            a_eq,
+            BooleanArray::from(vec![Some(true), None, Some(false)])
+        );
+    }
+    #[test]
+    fn test_eq_dyn_utf8_scalar() {
+        let array = StringArray::from(vec!["abc", "def", "xyz"]);
+        let array = Arc::new(array);
+        let a_eq = eq_dyn_utf8_scalar(array, "xyz").unwrap();
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:2813]: 🎉

**Review by alamb — COMMENTED** [review]
Thank you @matthewmturner -- I think this is looking great ❤️

**Comment by alamb** [comment]
How shall we proceed -- get this one polished up (maybe also add `eq_dyn_scalar_bool`?) and then start hammering on the other kernels (e.g. `neq`, `lt`, `gt`, etc?)

**Comment by matthewmturner** [comment]
> How shall we proceed -- get this one polished up (maybe also add `eq_dyn_scalar_bool`?) and then start hammering on the other kernels (e.g. `neq`, `lt`, `gt`, etc?)

Sure sounds good.  Should I just add those on this PR?

**Comment by matthewmturner** [comment]
actually - are `bool` values currently supported with `DictionaryArray`?

I get the following when trying to make one:

```
a value of type `array_dictionary::DictionaryArray<datatypes::types::Int8Type>` cannot be built from an iterator over elements of type `bool`
the trait `FromIterator<bool>` is not implemented for `array_dictionary::DictionaryArray<datatypes::types::Int8Type>`
```

i also dont see any tests for `DictionaryArray` that has that as value type.

**Comment by matthewmturner** [comment]
@alamb if you're okay with the latest changes and can merge then ill prioritize doing another PR with the other comparison functions.

thank you for all the guidance youve provided on this!

File: arrow/src/compute/kernels/comparison.rs [hunk 14]
```diff
@@ -898,6 +900,305 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+macro_rules! dyn_compare_scalar {
+    ($LEFT: expr, $RIGHT: expr, $OP: ident) => {{
+        let right: i128 = $RIGHT.try_into().map_err(|_| {
+            ArrowError::ComputeError(String::from("Can not convert scalar to i128"))
+        })?;
+        match $LEFT.data_type() {
+            DataType::Int8 => {
+                let right: i8 = right.try_into().map_err(|_| {
+                    ArrowError::ComputeError(String::from(
+                        "Can not convert scalar to i128",
```
Comment (liukun4515) [inline arrow/src/compute/kernels/comparison.rs:912]: `Can not convert scalar to i128` -> `Can not convert scalar to i8`

File: arrow/src/compute/kernels/comparison.rs [hunk 15]
```diff
@@ -1104,27 +1104,71 @@ where
...
-    };
+        _ => Err(ArrowError::ComputeError(
+            "Kernel only supports PrimitiveArray or DictionaryArray with Primitive values".to_string(),
+        ))
+    }
 }
 
 /// Perform `left == right` operation on an array and a numeric scalar
 /// value. Supports StringArrays, and DictionaryArrays that have string values
 pub fn eq_dyn_utf8_scalar(left: Arc<dyn Array>, right: &str) -> Result<BooleanArray> {
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:1138]: love it

File: arrow/src/compute/kernels/comparison.rs [hunk 16]
```diff
@@ -1104,27 +1104,71 @@ where
...
+/// Perform `left == right` operation on an array and a numeric scalar
+/// value. Supports BooleanArrays, and DictionaryArrays that have string values
+pub fn eq_dyn_bool_scalar(left: Arc<dyn Array>, right: bool) -> Result<BooleanArray> {
+    let result = match left.data_type() {
+        DataType::Boolean => {
+            let left = as_boolean_array(&left);
+            eq_bool_scalar(left, right)
+        }
+        _ => Err(ArrowError::ComputeError(
+            "Kernel only supports BooleanArray".to_string(),
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:1168]: Yeah -- users can call `cast` if they want to convert their array into boolean first

**Comment by alamb** [comment]
> Sure sounds good. Should I just add those on this PR?


I recommend doing the code in multiple PRs to keep the reviews smaller. 

> actually - are bool values currently supported with DictionaryArray?

I think they would be valid per the arrow spec, but I don't think they would be very useful -- A dictionary array of bools will be (much) less efficient both in term of space and CPU than a BooleanArray. This is probably why there is no array builder for them https://docs.rs/arrow/6.4.0/arrow/array/struct.DictionaryArray.html?search=dictionarybuilder 

I think it is fine to dictionary of bools unimplemented and people can implement them if they want

File: arrow/src/compute/kernels/comparison.rs (hunk 14 above)
Comment (alamb) [reply arrow/src/compute/kernels/comparison.rs:912]: It affects the messages below as well

File: arrow/src/compute/kernels/comparison.rs [hunk 17]
```diff
@@ -898,6 +900,305 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
...
+                })?;
+                let left = as_primitive_array::<UInt64Type>($LEFT);
+                $OP::<UInt64Type>(left, right)
+            }
+            _ => Err(ArrowError::ComputeError(String::from(
+                "Unsupported data type",
+            ))),
+        }
+    }};
+    ($LEFT: expr, $RIGHT: expr, $KT: ident, $OP: ident) => {{
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:986]:
```suggestion
    /// Applies `LEFT OP RIGHT` when `LEFT` is a `DictionaryArray` with keys of type `KT`
    ($LEFT: expr, $RIGHT: expr, $KT: ident, $OP: ident) => {{
```

File: arrow/src/compute/kernels/comparison.rs [hunk 18]
```diff
@@ -898,6 +900,305 @@ pub fn gt_eq_utf8_scalar<OffsetSize: StringOffsetSizeTrait>(
     compare_op_scalar!(left, right, |a, b| a >= b)
 }
 
+macro_rules! dyn_compare_scalar {
+    ($LEFT: expr, $RIGHT: expr, $OP: ident) => {{
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:904]:
```suggestion
    /// Applies `LEFT OP RIGHT` when `LEFT` is a `DictionaryArray` with keys of type `KT`
    ($LEFT: expr, $RIGHT: expr, $OP: ident) => {{
```

**Review by alamb — APPROVED** [review]
The code looks great to me. I had a few small suggestions, but otherwise all good 👍

**Comment by matthewmturner** [comment]
@alamb updates made!

**Comment by matthewmturner** [comment]
seems im getting hit with that `from_str_unchecked` issue

**Comment by alamb** [comment]
> seems im getting hit with that from_str_unchecked issue


I think it has been resolved if you merge up from master again

**Comment by matthewmturner** [comment]
having issues with rebasing. for whatever reason when i rebase to upstream master im losing all my recent commits.  still looking into it.

**Comment by matthewmturner** [comment]
after wandering through git hell for a bit i think i figured it out.  hopefully CI passes and were good here.

**Review by alamb — APPROVED** [review]
Looks great @matthewmturner  -- shall I merge this as is, or do you want to apply the suggestions from @liukun4515 first (fix error messages)?

Then, it seems like the next step might be to file tickets (I can do so if you wish) for the other kernels (`lt_dyn_scalar`, `gt_dyn_scalar`, etc)

**Comment by matthewmturner** [comment]
@alamb odd, i had made updates for those suggestions. must have been missed with my rebase.

ive made the fixes.

i can make an issue for it - and im hoping to work on implementing it this weekend.

do you have a preference for how small (i.e how many kernels) youd like each pr / issue to be?

**Comment by alamb** [comment]
> i can make an issue for it - and im hoping to work on implementing it this weekend.

Thanks @matthewmturner  ❤️ 


> do you have a preference for how small (i.e how many kernels) youd like each pr / issue to be?

My personal bias is one PR per kernel (as that makes it easiest to review) but I am also happy to review a single PR too

**Comment by liukun4515** [comment]
> > i can make an issue for it - and im hoping to work on implementing it this weekend.
> 
> Thanks @matthewmturner ❤️
> 
> > do you have a preference for how small (i.e how many kernels) youd like each pr / issue to be?
> 
> My personal bias is one PR per kernel (as that makes it easiest to review) but I am also happy to review a single PR too

I think one pr  per kernel is better.
The pr including all updates may be large, and it is not friendly to reviewers.
@matthewmturner

**Review by liukun4515 — APPROVED** [review]
LGTM

**Comment by matthewmturner** [comment]
@alamb @liukun4515 one PR per kernel it is :)

**Comment by alamb** [comment]
Thanks again @matthewmturner  and @liukun4515  -- this is great.
